Add fixed-size FIFO implementation

Mattemagikern opened the pull request:
This PR aims to introduce a fixed-size FIFO implementation in Zephyr's utils library. 
It is built on top of the existing ring buffer and abstracts item management away from both the ring buffer and the user.

The goal is to improve readability and reduce boilerplate in code that previously used ring buffers to implement FIFO behavior.

Mattemagikern commented:
Note to self: 
`subsys/net/lib/lwm2m/` seems to be a fixed size fifo as well, with some added logic to restore the cache.

Mattemagikern committed:
tests/lib: add test suite for ringq API

New test suite covering the fixed size ringq implementation.
The tests verify correct behavior for put(enqueue), get(dequeue),
wraparound, and overflow handling.
Note this data structure & tests relies on the correct behaviour of the
underlying ring buffer implementation.

Includes:
- Basic functional tests for ringq_put/ringq_get
- Boundary and wraparound cases
- Validation of error handling and empty/full conditions

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
utils: ringq: add fixed-size queue implementation

This commit introduces a fixed-size Queue implementation.
It is built on top of the existing ring buffer and abstracts item
management away from both the ring buffer and the user.

The goal is to improve readability and reduce boilerplate in code that
previously used ring buffers to implement FIFO behavior of items.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
drivers: i2s: sam_ssc: use sys/ringq.h instead of ring_buffer

The SSC I2S driver was using a custom ring buffer implementation for
buffering audio data. This commit replaces the custom ring_buffer
implementation with the sys_ringq implementation to improve
maintainability and code reuse.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
lib: utils: ring_buffer: deprecate ring_buffer item api

Deprecate the ring_buffer item API in favor of the `<zephyr/sys/ringq.h>`
to reduce maintenance and improve readability.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
drivers: console: switch from ring_buf_item to ring_buf

This commit updates the ipm_console_receiver driver to use the standard
ring_buf API instead of ring_buf_item*.

Since the ipm_console driver operates at the byte level, there is no need
to use the item-level API. This makes way for the removal of ring_buf_item
api in the future.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
doc: kernel: ring_buffer: Remove documentation for item api

Documentation for the item API for ring buffers has been removed due to it
has been deprecated and an alternative (<zephyr/sys/ringq.h>) has been
provided for similar functionality. This change aims to prevent future use
of the item API by obscuring it until it can be fully retired.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
doc: kernel: data_structures: sys_ringq

Adding documentation for the sys_ringq data structure.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
drivers: input: tsc_keys: Use fifo abstraction for input events

Replace the ring buffer–based event queue with the new sys_ringq API to
simplify input event handling. The sys_ringq abstraction removes the need
for manual item size management and improves readability by removing logic
regarding item boundries.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
drivers: i2s: litex: ring_buffer => ringq

Replacing the i2s_litex ring_buffer implementation with the
sys_ringq implementation to improve maintainability and code reuse.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
net: lwm2m: Use sys_ringq to store lwm2m_time_series_elem

The lwm2m registry time series elements were previously stored in a
ring buffer which added complexity to the code when adding and removing
elements from the buffer. This commit changes the implementation to use
the sys_ringq instead, simplifying the code and making it easier to
maintain over time.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
drivers: i2s: silabs_siwx91x: use sys_ringq instead of ring_buffer

Use the sys_ringq instead of the custom ring_buffer implementation for
SiWx91x I2S driver.
The aim is to reduce code duplication and improve maintainability.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern commented:
Please remove the stale label.

bjarki-andreasen commented on include/zephyr/sys/fifo.h:
```diff
@@ -0,0 +1,163 @@
+/**
+ * @brief A structure to represent a fifo queue
+ */
+struct fifo {
```
namespace with `sys_`

bjarki-andreasen reviewed (changes requested):
I like it, its very clean, and I particularly never liked the ring_buf_item APIs.

maass-hamburg commented on lib/utils/fifo.c:
```diff
@@ -0,0 +1,76 @@
+
+bool sys_fifo_full(const struct sys_fifo *fifo)
+{
+	return sys_fifo_space(fifo) == 0;
```
```suggestion
	return ring_buf_space_get(&fifo->rb) < fifo->item_size;
```
that way no division is needed

bjarki-andreasen reviewed (commented):
Nice!

npitre reviewed (commented):
Nice work on this FIFO abstraction — the API is clean and the driver conversions look good. A few suggestions below, mostly around an optimization opportunity in `put()`/`get()`/`full()` that also has implications for the `sys_fifo_init()` adjustment discussion.

npitre commented on include/zephyr/sys/fifo.h:
```diff
@@ -0,0 +1,227 @@
+	}
+
+	ring_buf_get(&fifo->rb, (uint8_t *)element, fifo->item_size);
+	return 0;
```
Same optimization as `sys_fifo_put()` — `ring_buf_get()` returns 0 when empty:
```c
static inline int sys_fifo_get(struct sys_fifo *fifo, void *element)
{
	if (ring_buf_get(&fifo->rb, (uint8_t *)element, fifo->item_size) == 0) {
		return -ENODATA;
	}
	return 0;
}
```

npitre commented on include/zephyr/sys/fifo.h:
```diff
@@ -0,0 +1,227 @@
+ */
+static inline bool sys_fifo_full(const struct sys_fifo *fifo)
+{
+	return ring_buf_space_get(&fifo->rb) < fifo->item_size;
```
With the item_size-multiple invariant, the free space is always either 0 or >= `item_size`. So `< fifo->item_size` is equivalent to `== 0`, which avoids loading `item_size` from memory:
```c
static inline bool sys_fifo_full(const struct sys_fifo *fifo)
{
	return ring_buf_space_get(&fifo->rb) == 0;
}
```

npitre commented on doc/kernel/data_structures/fifo.rst:
```diff
@@ -0,0 +1,88 @@
+
+.. code-block:: c
+
+    FIFO_INIT(my_fifo, item_size, item_capacity, buffer_pointer);
```
`FIFO_INIT()` doesn't exist. This should be `SYS_FIFO_DEFINE()` to match the actual macro:
```rst
    SYS_FIFO_DEFINE(my_fifo, item_size, item_capacity);
```
(Note: `SYS_FIFO_DEFINE` takes 3 args, not 4 — there's no `buffer_pointer` since the macro creates its own backing array.)

npitre replied:
Let's just call that thoroughness a form of teamwork 😉

npitre commented on include/zephyr/sys/fifo.h:
```diff
@@ -0,0 +1,227 @@
+	}
+
+	ring_buf_put(&fifo->rb, (const uint8_t *)element, fifo->item_size);
+	return 0;
```
Actually, this can be a one-liner:
```c
static inline int sys_fifo_put(struct sys_fifo *fifo, const void *element)
{
	return ring_buf_put(&fifo->rb, (const uint8_t *)element, fifo->item_size) == 0
		? -ENOSPC : 0;
}
```
And similarly for `sys_fifo_get()`.

npitre commented on include/zephyr/sys/fifo.h:
```diff
@@ -0,0 +1,227 @@
+	__ASSERT(data_size % item_size == 0, "Data size should be a multiple of item size");
+
+	/* Adjust data size to be a multiple of item size */
+	data_size = (data_size / item_size) * item_size;
```
Fair point. The alignment assert should go since rounding down makes it a legal operation. But asserts for `item_size \!= 0` and `data_size \!= 0` should be added (div-by-zero). And document that `data_size` gets rounded down.

Mattemagikern replied:
Thank you both for your reviews & the optimization pointers! :) 

I've added the changes to the latest patch-set, as @nashif pointed out there may be of interest to rename the API so I uploaded the initial changes for your consideration until that discussion has been resolved :)

npitre commented on doc/kernel/data_structures/fifo.rst:
```diff
@@ -0,0 +1,88 @@
+.. _fixed_size_fifo_api:
+
+Fifo Data Structure
```
There is precedent for this naming: `sys_heap` provides a lockless allocator and `k_heap` adds kernel synchronization on top. `k_fifo` is not built on top of `sys_fifo` (it wraps `k_queue`, a linked-list), but the similarity in functionality — both are FIFO queues, one lockless, one synchronized — could justify the similarity in name following the same `sys_` / `k_` layering pattern.

Mattemagikern replied:
@andyross @peter-mitsis @maass-hamburg @nashif @npitre 

Friendly reminder :)

cfriedt reviewed (commented):
LGTM - just one recommendation to drop `SYS_FIFO_DEFINE_TYPED()`

nashif reviewed (changes requested):
see https://github.com/zephyrproject-rtos/zephyr/pull/98255#discussion_r2872629609

andyross reviewed (dismissed):
Late to the party.  The core implementation looks clean to me.  I take no position on the naming bikeshed (though I do point out that "ringq" has the advantage of being short).

mathieuchopstm commented on include/zephyr/sys/ring_buffer.h:
Missing words in commit message for `lib: utils: ring_buffer: deprecate ring_buffer item api`?

> in favor of the reduce maintenance and improve readability

Mattemagikern replied:
Haha, that is very correct. Should be fixed! :)

mathieuchopstm commented on doc/kernel/data_structures/ring_buffers.rst:
In addition to `doc: kernel: ring_buffer: Remove documentation for item api`, an entry in the migration guide for v4.5 should be added.

mathieuchopstm commented on doc/kernel/data_structures/ringq.rst:
```diff
@@ -0,0 +1,83 @@
+
+The ringq is used whenever you need to decouple producers and consumers of discrete
+data sizes without worrying about partial reads or variable-length payloads. This
+differentiation separates it from the ring_buffer data structure, which is a streaming data
```
Perhaps make `ring_buffer` a hyperlink to the Ring Buffer data structure page?

Mattemagikern replied:
Thanks for the reviews @mathieuchopstm ! :)

Mattemagikern committed:
doc: releases: 4.5: note ring_buf item API deprecation

Add a migration-guide entry for the ring_buf item API deprecation, and add
release-notes entries under "Deprecated APIs" and "New APIs" describing
the deprecation of the ring_buf item API and the introduction of
sys_ringq as its replacement.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>
